routing/router.go

```diff
@@ -155,8 +155,12 @@ func serveEnvJsFile(w http.ResponseWriter) {
 		body += fmt.Sprintf("window.HISTORY_MODE='%s';", conf.Server.WebHistoryMode)
 	}
 
-	if webRoot := strings.TrimSuffix(conf.Server.WebRoot, "/"); len(webRoot) > 0 {
-		body += fmt.Sprintf("window.WEB_ROOT='%s';", webRoot)
+	if redirectPath := conf.Server.RedirectPath; len(redirectPath) == 0 {
+		if webRoot := strings.TrimSuffix(config.Get().Server.WebRoot, "/"); len(webRoot) > 0 {
+			body += fmt.Sprintf("window.WEB_ROOT='%s';", webRoot)
+		}
+	} else {
+		body += fmt.Sprintf("window.WEB_ROOT='%s';", redirectPath)
```

Comment:
I was checking resulting outputs whe `kubectl proxy` is in use. We also have this `serveIndexFile` function which also prints the `web_root` setting in the index HTML file that is served:

https://github.com/kiali/kiali/blob/f444a09521033b1f9fbc5161e2ce65dc0c830001/routing/router.go#L171-L197

I see that `kubectl proxy` is doing the right URL rewrites in the resulting HTML. So, instead of adding yet another config setting, why not replace all this with something that does not depends on configs? So, in this `serveEnvJsFile` function, I think something like this would work:

```javascript
body += "window.WEB_ROOT = document.getElementsByTagName('base')[0].getAttribute('href')"
```

The previous line would completely replace the `if` block in this `serveEnvJsFile` function.
This way we will no longer need to add that `redirect_path` option and we would completely rely on the proxy url_rewrite logic.

I haven't tested, but I think this simpler change would be better, and it would save us from adding, and documenting, and explaining a new option, and how the new option interacts with other settings.

Reply:
> So, in this serveEnvJsFile function, I think something like this would work:
> body += "window.WEB_ROOT = document.getElementsByTagName('base')[0].getAttribute('href').replace(/\/+$/g, '')"
> The previous line would completely replace the if block in this serveEnvJsFile function.

This is an interesting alternative solution. Let's try to see if that would work - as you say, it means we won't have to do anything with implementing/documenting a new config setting.

@doflamingo721 see if you can try that solution. It would require a single PR with a single line change apparently :)

But then we have to examine this - but I assume nothing has to change here: https://github.com/kiali/kiali/blob/f444a09521033b1f9fbc5161e2ce65dc0c830001/routing/router.go#L171-L197

Reply:
@israel-hdez @doflamingo721 I tried that alternative solution and it does not work. Here's the env.js:

```
$ curl -L http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali/env.js && echo
window.WEB_ROOT = document.getElementsByTagName('base')[0].getAttribute('href').replace(//+$/g, '')
```

and accessing the UI at http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali produces the same error as before - we get redirected to a URL that does not include the proxy (I see it going to this: http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali). So that alternative is not viable.

Reply:
My fault, but probably the regex is badly escaped. I see I posted `//+$/g `. Probably it should be `/\/+$/g ` (notice the escaped slash).

If in doubt, you probably can try that whole JS expression `document.get.....` in the browser console while you are in Kiali to see if it gets the right URL

Reply:
What I mean is that in the resulting env.js you should see `/\/+$/g `. Probably golang is doing some escaping on its side.

Reply:
oh, look at that.. it works :)

```
$ curl -L http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali/env.js && echo
window.WEB_ROOT = document.getElementsByTagName('base')[0].getAttribute('href').replace(/\/+$/g, '')
```

Going to http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali
brings the browser to: http://localhost:8080/api/v1/namespaces/istio-system/services/kiali:20001/proxy/kiali/console/overview?duration=60&refresh=15000

The diff:
```diff
diff --git a/routing/router.go b/routing/router.go
index ba6828b51..6c46341bb 100644
--- a/routing/router.go
+++ b/routing/router.go
@@ -155,9 +155,7 @@ func serveEnvJsFile(w http.ResponseWriter) {
                body += fmt.Sprintf("window.HISTORY_MODE='%s';", conf.Server.WebHistoryMode)
        }
 
-       if webRoot := strings.TrimSuffix(conf.Server.WebRoot, "/"); len(webRoot) > 0 {
-               body += fmt.Sprintf("window.WEB_ROOT='%s';", webRoot)
-       }
+       body += "window.WEB_ROOT = document.getElementsByTagName('base')[0].getAttribute('href').replace(/\\/+$/g, '')"
```

Reply:
I did a little more testing, and that small alternative solution appears to be good. @doflamingo721 barring any errors you see in your official setup with this new way of implementing this, we should do that instead.

The only difference I see is it will affect everyone (in this original solution, the behavior is identical as before; only if you set redirect_path will window.WEB_ROOT be different than before). However, I don't see a problem right now. I'll keep playing with it to see if I can break it, but so far it seems to be doing what we want.

Reply:
@jmazzitelli @israel-hdez Thank you for your inputs on this. I'll try out the alternative solution for this at the earliest and if everything seems to workout I'll update my PR and re-request your reviews.

Reply:
@doflamingo721 please test out that alternative solution and let us know. All tests pass over here.

Reply:
Hey @jmazzitelli @israel-hdez , The alternative solution worked for me on my minikube setup. However the same was not working on my rancher setup. The `window.WEB_ROOT` was getting set to the entire browser path including the protocol and domain name. So the browser window was getting rewritten as `https://128.xx.xx.xx/https://128.xx.xx.xx/path/to/kiali-service/kiali:20001/proxy`. To fix this I modified the regex and it is working fine on my setups. Can you please review this change and let me know if it looks good to you?

Reply:
Right, the regex originally proposed is basically a trimEnd of the `/` slash. It is done with a regex because, AFAIK, the `trimEnd` func in Javascript only works with blanks (it does not receives what character you want to trim).